src/app.rs

```diff
@@ -32,6 +34,12 @@ pub struct App {
     /// The server configuration
     pub config: config::Server,
 
+    /// Cache the `version_id` of a `canonical_crate_name:semver` pair
+    ///
+    /// This is used by the download endpoint to reduce the number of database queries. The
+    /// `version_id` is only cached under the canonical spelling of the crate name.
+    pub(crate) version_id_cacher: DashMap<String, i32>,
```

Comment:
We can support serving responses for non-canonical spellings with little effort:

```rust
struct CachedVersion {
    canonical_crate_name: String,
    version_id: i32,
}

version_id_cacher: DashMap<String, CachedVersion>
```

Then the cache key could be normalized with `crate_name.lower().replace('_', '-')`, while still having access to the canonical crate name to perform the redirect.

Reply:
I was originally storing the canonical crate name as well, but ended up dropping that to minimize memory usage, especially since there are so few non-canonical requests.

I'm not actually that worried about memory usage, so this approach does sound reasonable to me as well. Maybe it's worth looking at memory usage in production first, and then adding this if we're sure we have plenty of room to spare.